src/System.Threading.Channels/src/System/Threading/Channels/ChannelReader.netcoreapp.cs

```diff
@@ -0,0 +1,26 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace System.Threading.Channels
+{
+    public abstract partial class ChannelReader<T> : IAsyncEnumerable<T>
+    {
+        IAsyncEnumerator<T> IAsyncEnumerable<T>.GetAsyncEnumerator() => AsAsyncEnumerable().GetEnumerator();
+
+        public virtual async IAsyncEnumerable<T> AsAsyncEnumerable(CancellationToken cancellationToken = default)
+        {
+            while (await WaitToReadAsync(cancellationToken).ConfigureAwait(false))
+            {
+                while (TryRead(out T item))
+                {
+                    yield return item;
+                    cancellationToken.ThrowIfCancellationRequested();
```

Comment:
>cancellationToken.ThrowIfCancellationRequested(); [](start = 20, length = 49)

Is it possible we get into the state that TryRead(out T item) succeed and then we do yield return item; but after this point the the cancellation happen. at that time I don't think we need to throw as next TryRead(out T item) should return false and we exit the loop.

in other word, the channel becomes empty and then the cancellation happen.

Reply:
> the channel becomes empty and then the cancellation happen.

That could be the case no matter where we check.

- If we check before we enter the outer loop, then we won't be checking per item.
- If we check before we enter the inner loop, someone else could have already taken out the item that caused WaitForReadAsync to return true, and then the channel could be marked as completed and TryRead will return false.
- If we check after the TryRead but before the yield we might consume an item and then throw, losing data.
- If we check after the yield, then we're at least checking for each item, but may throw on cancellation even if the channel is complete.  We could of course try to check for that, too, e.g. `if (token.IsCancellationRequested && !Completed) cancellationToken.ThrowIfCancellationRequested();`, but there's still a race condition there.